fedimint-client/src/module/mod.rs

```diff
@@ -183,6 +184,57 @@ pub trait ClientModule: Debug + MaybeSend + MaybeSync + 'static {
     async fn subscribe_balance_changes(&self) -> BoxStream<'static, ()> {
         unimplemented!()
     }
+
+    /// Leave the federation
+    ///
+    /// While technically there's nothing stopping the client from just
+    /// abandoning Federation at any point by deleting all the related
+    /// local data, it is useful to make sure it's safe beforehand.
+    ///
+    /// This call indicates the desire of the caller client code
+    /// to orderly and safely leave the Federation by this module instance.
+    ///
+    /// The client module implementation should do whatever necessary and
+    /// possible to serve this request. E.g. persist this
+    /// state, preventing new interactions with the Federation. Notably
+    /// modules should not disable interaction that might be necessary
+    /// for the user (possibly through other modules) to leave. In particular
+    /// a Mint module should retain ability to create new modules,
+    /// and LN module should retain ability to send funds out.
+    ///
+    /// This function should return an error if it's not currently possible
+    /// to safely (e.g. without loosing funds) leave the Federation.
+    /// It should avoid running indefinitely trying to complete any cleanup
+    /// actions necessary to reach a clean state, preferring spawning new
+    /// state machines and returning an informative error.
+    ///
+    /// If any action/change of state is required to leave, this function
+    /// MUST return a `Err`.
+    ///
+    /// This function should return `Ok` only if from the perspective
+    /// of this module instance, it is safe to delete its data and
+    /// stop using it.
+    ///
+    /// Calling code must NOT assume that a module that once returned `Ok`,
+    /// will not return `Err` at later point. E.g. a Mint module might have
+    /// no outstanding balance at first, but other modules winding down
+    /// might "cash-out" to Ecash.
+    ///
+    /// Before leaving the Federation and deleting any state the calling code
+    /// must collect a full round of `Err`.
+    ///
+    /// Calling code should allow the user to override and ignore any
+    /// outstanding errors, after sufficient amount of warnings. Ideally,
+    /// this should be done on per-module basis, to avoid mistakes.
+    async fn leave(
```

Comment:
Is it correct to say that this function shouldn't have any side effects. Meaning it should simply observe the current state and report back whether its safe to leave or not depending on that state (without changing any state)?
- If the above is correct, is there a way to enforce that through code? Perhaps somehow making the DB handle read-only?
- If the above is correct, how/where do we envision doing the necessary state changes to actually be able to leave? And what if those state changes take several minutes?

Reply:
I think it should have side-effects. Like closing down pending contracts, etc.

Reply:
Doc is vague in that case:

```
    /// The client module implementation should do whatever necessary and
    /// possible to serve this request. E.g. persist this
    /// state, preventing new interactions with the Federation.
```

^^ this makes it seem like module should do whatever is necessary _within the leave() function_.

```
    /// It should avoid running indefinitely trying to complete any cleanup
    /// actions necessary to reach a clean state, preferring spawning new
    /// state machines and returning an informative error.
```

^^ this makes it seem like _actually maybe we should think twice about doing whatever is necessary_. We might think doing a contract clean up or freeing up some locked balance should only take 30s but what if it could take 10+ minutes in some cases? As a module developer reading this documentation leaves me a bit confused.

It would be really nice to resolve ambiguity. Something more opinionated but specific would be nice, even if it was half the amount of documentation. Maybe even including an example in the documentation, like

```
    // Let's say there is a pending lightning contract which cannot be
    // cancelled and we have to wait X blocks for it to expire or go through.
    // In the case this function should do X, Y, and finally Z.
```